### PR TITLE
Modified SANA core frequency scoring

### PR DESCRIPTION
**Pull request by rasulsafa**

### CODE REVIEW

**Review by rasulsafa — COMMENTED** [review]
Tweaked core frequency measurements

File: src/methods/SANA.cpp [hunk 1]
```diff
@@ -315,7 +315,7 @@ Alignment SANA::run() {
             align = hillClimbingAlignment(align, (long long int)(10000000)); //arbitrarily chosen, probably too big.
             cout << hill.elapsedString() << endl;
         }
-#define PRINT_CORES 0
+#define PRINT_CORES 1
```
Comment (waynebhayes) [inline src/methods/SANA.cpp:318]: Leave PRINT_CORES 0 by default, and add code instead to put these scores (unweighted only for now) to a ".scores" file.

File: src/methods/SANA.cpp [hunk 2]
```diff
@@ -1106,9 +1106,37 @@ void SANA::performChange(int type) {
             item.second += localScoreSumIncChangeOp(localSimMatrixMap[item.first], source, oldTarget, newTarget);
     }
 
+
+
     double newCurrentScore = 0;
     bool makeChange = scoreComparison(newAligEdges, newInducedEdges, newTCSum, newLocalScoreSum, newWecSum, newNcSum, newCurrentScore, newEwecSum, newSquaredAligEdges, newExposedEdgesNumer);
+
+    if(wasBadMove && !makeChange)
```
Comment (waynebhayes) [inline src/methods/SANA.cpp]: I think the only thing we want to find out is which hole is the better place to be; set a variable equal to either source or newTarget; and then have the code inside the #if CORES only once, using that variable.

File: src/methods/SANA.cpp [hunk 3]
```diff
@@ -1106,9 +1106,37 @@ void SANA::performChange(int type) {
...
+			static double pBad;
+			if(trueAcceptingProbability()>0) pBad = trueAcceptingProbability();
+			coreCount[source]++;
+			coreFreq[source][(*A)[source]]++;
+			weightedCoreFreq[source][(*A)[source]] += pBad;
+			totalCoreWeight[source] += pBad;
+		#endif
+    }
+
     if (makeChange) {
```
Comment (waynebhayes) [inline src/methods/SANA.cpp:1126]: Consider the case that it's a bad move but we're going to accept it anyway. Then the above (wasBadMove && !makeChange) is false and doesn't get executed. Then here, where the bad move is accepted, you're incrementing the bad guy's core count.

**Review by waynebhayes — CHANGES_REQUESTED** [review]

File: src/methods/SANA.cpp [hunk 4]
```diff
@@ -1111,31 +1111,19 @@ void SANA::performChange(int type) {
...
-			coreCount[source]++;
-			coreFreq[source][(*A)[source]]++;
-			weightedCoreFreq[source][(*A)[source]] += pBad;
-			totalCoreWeight[source] += pBad;
-		#endif
-    }
+    #ifdef CORES
+		// Statistics on the emerging core alignment.
+		// only update pBad if it's nonzero; re-use previous nonzero pBad if the current one is zero.
+        uint betterHole = wasBadMove ? source : newTargetIndex;
```
Comment (waynebhayes) [inline src/methods/SANA.cpp:1117]: Good idea, but I think also if you create a "worseHole" variable it'll be useful below. See later comment.

File: src/methods/SANA.cpp [hunk 5]
```diff
@@ -1208,41 +1196,34 @@ void SANA::performSwap(int type) {
...
+        weightedCoreFreq[source1][(*A)[source1]] += pBad;
+        totalCoreWeight[source1] += pBad;
 
-		#endif
-    }
+        coreCount[source2]++;
+        weightedCoreFreq[source2][(*A)[source2]] += pBad;
+        totalCoreWeight[source2] += pBad;
 
+        if(wasBadMove)
```
Comment (waynebhayes) [inline src/methods/SANA.cpp]: If you create the "worseHole" above, then you can get rid of this if-else statement and just use "bestHole" and "worseHole".

**Review by waynebhayes — APPROVED** [review]

**Pull request merged**